internal/controller/serverset/serverset_crossplane_mock_test.go

```diff
@@ -0,0 +1,242 @@
+		nicController        kubeNicControlManager
+		serverController     kubeServerControlManager
+		log                  logging.Logger
+	}
+	type args struct {
+		ctx context.Context
+		cr  *v1alpha1.ServerSet
+	}
+
+	tests := []struct {
+		name    string
+		fields  fields
+		args    args
+		want    managed.ExternalCreation
+		wantErr error
+	}{
+		{
+			name: "server set successfully created",
+			fields: fields{
+				kube: &test.MockClient{
+					MockList: test.NewMockListFn(nil, fakeObjList),
+				},
+				log: logging.NewNopLogger(),
+				mg:  &fake.Managed{},
+				bootVolumeController: &kubeBootVolumeController{
+					kube: &test.MockClient{
+						MockCreate: test.NewMockCreateFn(nil),
+						MockList:   test.NewMockListFn(nil, fakeObjList),
+						MockGet:    test.NewMockGetFn(nil, fakeGetObj),
+					},
+					log: logging.NewNopLogger(),
+				},
+				serverController: &kubeServerController{
+					kube: &test.MockClient{
+						MockCreate: test.NewMockCreateFn(nil),
+						MockList:   test.NewMockListFn(nil, fakeObjList),
+						MockGet:    test.NewMockGetFn(nil, fakeGetObj),
+					},
+					log: logging.NewNopLogger(),
+				},
+				nicController: &kubeNicController{
+					kube: &test.MockClient{
+						MockCreate: test.NewMockCreateFn(nil),
+						MockList:   test.NewMockListFn(nil, fakeObjList),
+						MockGet:    test.NewMockGetFn(nil, fakeGetObj),
+					},
+					log: logging.NewNopLogger(),
+				},
+			},
+			args: args{
+				ctx: context.Background(),
+				cr:  createBasicServerSet(),
+			},
+			want: managed.ExternalCreation{
+				ConnectionDetails: managed.ConnectionDetails{},
+			},
+			wantErr: nil,
+		},
+		{
+			name: "failed to create, too many volumes",
+			fields: fields{
+				kube: &test.MockClient{
+					MockList: test.NewMockListFn(nil, func(obj client.ObjectList) error {
+						vols := obj.(*v1alpha1.VolumeList)
+						vols.Items = append(append(vols.Items, v1alpha1.Volume{}), v1alpha1.Volume{})
+						return nil
+					}),
+				},
+				log: logging.NewNopLogger(),
+				mg:  &fake.Managed{},
+			},
+			args: args{
+				ctx: context.Background(),
+				cr:  createBasicServerSet(),
+			},
+			want: managed.ExternalCreation{
+				ConnectionDetails: nil,
+			},
+			wantErr: errors.New("found too many volumes for index 0"),
+		},
+		{
+			name: "external create failed",
+			fields: fields{
+				kube: &test.MockClient{
+					MockList: test.NewMockListFn(nil, fakeObjList),
+				},
+				log: logging.NewNopLogger(),
+				mg:  &fake.Managed{},
+				bootVolumeController: &kubeBootVolumeController{
+					kube: &test.MockClient{
+						MockCreate: test.NewMockCreateFn(nil),
+						MockList:   test.NewMockListFn(nil, fakeObjList),
+						MockGet: test.NewMockGetFn(nil, func(obj client.Object) error {
+							volume := obj.(*v1alpha1.Volume)
+							annotations := make(map[string]string)
+							annotations[meta.AnnotationKeyExternalCreateFailed] = "alas, we failed"
+							volume.SetAnnotations(annotations)
+							return nil
```

Comment:
You could extract the anonymous function into a defined function and re-use it in the next scenarios